SubApp separation and route protection

andioneto opened the pull request:
# **DOTORG-76: Separate and protect admin/user/public routes in react-router**
 
## Created 3 container/components to manage views by authentication permissions:
### **Current Contents:**
[note: L=> list view, V => single view, C => create]
- Public: login/logout, registration (localhoast:3000/public)
- Vounteers: landing-page (soon to be the active campaign list), account page, edit accout (localhost:3000/volunteers)
- Admin: account/account edit, campaigns(L, V, C), scripts(L, V, C), questions(L, V, C), contactLists(L)

**note:** Separated is_admin and logged in logic away from the store, currently on localStorage (TODO: utilize cookies?)
### **Current flow:**
- localhost:3000/ =()=> localhost:3000/admin =(is_admin + loggedIn)=> localhost:3000/admin/campaigns [from here all other admin views are accessible)
- localhost:3000/ =()=> localhost:3000/admin =(!is_admin + loggedIn)=> localhost:3000/volunteers/landingpage [from here all other volunteers views are accessible)
- localhost:3000/ =()=> localhost:3000/admin =(!is_admin + !loggedIn)=> localhost:3000/public/login [from here registration is also available)

**Change to other files:**
Had to make adjustments to header, nav, app, landing page, and any other page that had hard coded endpoints in Link or button elements.
Modified tests and updated snapshots.

andioneto committed:
begining of refactor, outline for admin routing component

andioneto committed:
made changes to: login action/reducer, headers/nav, auth controller to return is_Admin, backbone of subapp routing

andioneto committed:
added boolean for admin status and user id to local storage to deal with loss of store on refresh

bug: while logged in as a non-admin, if you open a new tab and try to navigate to localhost:3000, it gets stuck at the admin container.

andioneto committed:
updated FE tests after modifying subapp and auth separations

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into aodn-admin-user-public-sep

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into aodn-admin-user-public-sep

perfectlynormalbeast commented on public/src/components/header.jsx:
```diff
@@ -9,40 +9,51 @@ export default class Header extends Component {
         { title: 'Logout', href: '/logout' }
       ];
-      if (this.props.userInfo && this.props.userInfo.is_admin) {
+      if (isAdmin) {
```
Let's just drop this link here and rely on logo redirection?

andioneto replied:
So remove "All Campaigns" from the Dropdown?

perfectlynormalbeast commented on public/src/components/header.jsx:
```diff
@@ -9,40 +9,51 @@ export default class Header extends Component {
     if (userId) { // user is logged in aka id present
       links = [
-        { title: 'Account', href: `/account/${userId}` },
+        { title: 'Account', href: `${parent}/account/${userId}` },
```
This will always be `/users/account/${userId}` or `/users/account`?

andioneto replied:
Should always have the id in the endpoint, I think we had meant to add '/:id' into the router endpoint a while ago and forgot about it. I just did that.

JadziaHax commented:
I don't think this is functioning correctly.

If I'm not logged in and I go to localhost:3000 it forwards me to /admin and stops there.

I think the way the code looks I should be forwarded to /volunteers where it should check if I'm logged in, and then forward me to /public where I should be forwarded to /public/login, correct?

It does take me through one step each time I refresh the page after landing on /admin though.

perfectlynormalbeast commented on public/src/index.jsx:
```diff
@@ -5,143 +5,57 @@ import { Provider } from 'react-redux';
+            path="/public"
+            component={PublicContainer}
+          />
+          <Route
```
Can we drop this route and pull down the topmost route into a catchall route to redirect to /admin? This way any unmatched routes will be cascaded down the chain of routes.

andioneto replied:
Done!

perfectlynormalbeast commented on public/src/containers/sub_apps/volunteers_container.jsx:
```diff
@@ -0,0 +1,18 @@
+import { withRouter } from 'react-router-dom';
+import { connect } from 'react-redux';
+
+import Volunteers from '../../components/sub_apps/volunteers';
```
volunteers -> users in this file and filename?

andioneto replied:
Again, I changed this bc it was hitting the sever endpoint.

perfectlynormalbeast commented on public/src/components/sub_apps/volunteers.jsx:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  render() {
+    const parent = '/volunteers';
```
volunteers -> users in this file and filename?

andioneto replied:
Yeah, bc users was hitting our server

perfectlynormalbeast commented on public/src/components/sub_apps/volunteers.jsx:
```diff
@@ -0,0 +1,37 @@
+          path={`${parent}/account`}
+          component={ConnectedAccountPage}
+        />
+        <Route
```
Can we drop the path so it's a catchall?

andioneto replied:
done!

perfectlynormalbeast commented on public/src/components/sub_apps/public.jsx:
```diff
@@ -0,0 +1,30 @@
+        component={RegistrationForm}
+      />
+      <Route
+        path={`${parent}/login`}
```
What if we drop the path so it becomes the catchall? This way we can eliminate the topmost route too?

andioneto replied:
Ok, I did this for all of the route components. Let me know what you think.

perfectlynormalbeast commented on public/src/components/sub_apps/admin.jsx:
```diff
@@ -0,0 +1,84 @@
+        />
+        <Route
+          exact
+          path={`${parent}/account/:id/edit`}
```
I think this route should live under /users/account/:id/edit for now. Best not to connect a view under 2 different routes? Same goes for the view account page below this one?

andioneto replied:
Yeah, it's working for now, I wasn't sure if we were going to have an account view for admin. I can remove this if you think it's not needed here.

perfectlynormalbeast commented on public/src/components/sub_apps/admin.jsx:
```diff
@@ -0,0 +1,84 @@
+        />
+        <Route
+          exact
+          path={`${parent}`}
```
If we drop this path, it becomes a catchall? How do we feel about that?

andioneto replied:
What do you mean drop the path? Remove this routing logic?

perfectlynormalbeast commented on public/src/actions/login.jsx:
```diff
@@ -57,15 +63,23 @@ export function logoutUser() {
-  const token = localStorage.getItem('auth_token');
-  return !!id && !!token;
-}
+// DISCUSS: proposed removal
```
Seems fine to me.

perfectlynormalbeast reviewed (changes requested):
- Let's be less specific with `volunteers` and call it `users`? This will apply in multiple places and filenames possibly.
Some other comments

andioneto committed:
the routing was getting caught in the App component bc App is using this.props.children to render the rest of the application over App.
I dropped routes to let baselines be catchalls
Added logic to App component's componentDidMount method to replace the route with the appropirate sub app --> no more getting stuck at App and waving between routing containers!

andioneto committed:
removed extraneous Redrict componet import from admin file

andioneto committed:
updated App component tests, now passing

andioneto committed:
modified catch alls to be redirects

andioneto committed:
removed rogue console log

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into aodn-admin-user-public-sep

andioneto committed:
fixed redirect after registration to relfect updated routes

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into aodn-admin-user-public-sep

andioneto committed:
fixed registration action test to reflect changes in redirect to login

andioneto committed:
Merge branch 'master' into aodn-admin-user-public-sep

andioneto committed:
Merge branch 'master' into aodn-admin-user-public-sep
